### PR TITLE
Add HTML Renderer

### LINKED ISSUE #461: Create HTML renderer
In some parts of UI package, multiple widgets are used to display single sentences because parts of the sentence need to be formatted or styled. Therefore, a HTML renderer is needed for sentences that need formatting.

### PR DESCRIPTION
**Pull request by hydra0210**
fixes #461

### CODE REVIEW

**Review by beansgum — CHANGES_REQUESTED** [review]
- No text is rendered if I pass a text without html tags.


- Rendered text overflows the constrained width, the text should be wrapped to fit parent container.


This is how the dialog displays when using regular text label `th.Body1`.

File: ui/renderers/html.go [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+		doc:   doc,
+		theme: theme,
+	}
+	renderer.render()
+	return renderer
+}
+
+func (r *HTMLRenderer) render() {
+	r.doc.Find("*").Each(func(_ int, node *goquery.Selection) {
+		nodeName := goquery.NodeName(node)
```
Comment (beansgum) [inline ui/renderers/html.go]: It looks every single element is rendered separately which is what resulted in this. Rendering should be done in a nested way.
```html
    <span>Tap on <span style="text-color: primary">blue text</span> to copy <strong>the</strong> item.</span>
```

File: ui/renderers/html.go [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+		switch nodeName {
+		case "html", "head":
```
Comment (beansgum) [inline ui/renderers/html.go]: This is missing some basic html tags like bold `<b>`, italics `<i>`, font `<font>` & line-break `<br>`.

**Review by beansgum — CHANGES_REQUESTED** [review]
- Trying to render
```html
    Tap on <span style="text-color: primary">blue text</span> to copy the item.
```
results in:

File: ui/renderers/renderer.go [hunk 3]
```diff
@@ -7,34 +7,29 @@ import (
 	"unicode"
 
 	"gioui.org/layout"
+	"gioui.org/text"
 	"gioui.org/widget"
 	"gioui.org/widget/material"
 
-	md "github.com/gomarkdown/markdown"
+	//md "github.com/gomarkdown/markdown"
```
Comment (beansgum) [inline ui/renderers/renderer.go]: cleanup imports

File: ui/modal_templates.go [hunk 4]
```diff
@@ -22,23 +23,30 @@ const (
...
 		func(gtx C) D {
-			text := th.Body1("After you or your counterparty has genrated a signature, you can use this form to verify the" +
-				" validity of the  signature. \n \nOnce you have entered the address, the message and the corresponding " +
-				"signature, you will see VALID if the signature appropriately matches the address and message, otherwise INVALID.")
-			text.Color = th.Color.Gray
-			return text.Layout(gtx)
+			text := `<div style="text-color: success; font-weight: bold">
+						After you or your counterparty has genrated a signature, you can use this form to verify the
+				 		validity of the  signature.
+				 		<span style="text-color: primary; font-weight: bold">Nested span</span>
```
Comment (beansgum) [inline ui/modal_templates.go]: There's unrelated text in this html.

File: ui/modal_templates.go [hunk 5]
```diff
@@ -81,45 +89,22 @@ func privacyInfo(th *decredmaterial.Theme) []layout.Widget {
 }
 
 func setupMixerInfo(th *decredmaterial.Theme) []layout.Widget {
+	text := `<div style="text-color: gray">
```
Comment (beansgum) [inline ui/modal_templates.go]: Use `span` for styling, `div` is not required anywhere in here.

File: ui/modal_templates.go [hunk 6]
```diff
@@ -22,23 +23,30 @@ const (
 func verifyMessageInfo(th *decredmaterial.Theme) []layout.Widget {
 	return []layout.Widget{
 		func(gtx C) D {
-			text := th.Body1("After you or your counterparty has genrated a signature, you can use this form to verify the" +
-				" validity of the  signature. \n \nOnce you have entered the address, the message and the corresponding " +
-				"signature, you will see VALID if the signature appropriately matches the address and message, otherwise INVALID.")
-			text.Color = th.Color.Gray
-			return text.Layout(gtx)
+			text := `<div style="text-color: success; font-weight: bold">
```
Comment (beansgum) [inline ui/modal_templates.go]: The styling here is off, reference the design mockup for the proper text style.

File: ui/modal_templates.go [hunk 7]
```diff
@@ -81,45 +89,22 @@ func privacyInfo(th *decredmaterial.Theme) []layout.Widget {
 }
 
 func setupMixerInfo(th *decredmaterial.Theme) []layout.Widget {
+	text := `<div style="text-color: gray">
+				Two dedicated accounts (“mixed” & “unmixed”) will be created in order to use the mixer.
+			</div>
+			<h4>This action cannot be undone.</h4>`
```
Comment (beansgum) [inline ui/modal_templates.go]: This text is actually bold and it should inherit the text size of the parent tag which means it should be inside the parent tag. Reference the design mockup for proper text styling.
```suggestion
			<b>This action cannot be undone.</b>`
```

File: ui/modal_templates.go [hunk 8]
```diff
@@ -81,45 +89,22 @@ func privacyInfo(th *decredmaterial.Theme) []layout.Widget {
...
-		},
+		renderers.RenderHTML(text, th).Layout,
 	}
 }
 
 func transactionDetailsInfo(th *decredmaterial.Theme) []layout.Widget {
+	text := `<span style="text-color: gray">Tap on</span>
+			<span style="text-color: primary">blue text</span>
+			<span style="text-color: gray">to copy the item</span>`
+
```
Comment (beansgum) [inline ui/modal_templates.go]: A sentence like this should be render as:
```html
<span style="text-color: gray">Tap on <span style="text-color: primary">blue text</span> to copy the item</span>
```

File: ui/modal_templates.go [hunk 9]
```diff
@@ -81,45 +89,22 @@ func privacyInfo(th *decredmaterial.Theme) []layout.Widget {
...
-					})
-				}),
-				layout.Rigid(func(gtx C) D {
-					t := th.Body1("to copy the item.")
-					t.Color = th.Color.Gray
-					return t.Layout(gtx)
-				}),
-			)
-		},
+		renderers.RenderHTML(text, th).Layout,
```
Comment (beansgum) [inline ui/modal_templates.go:102]: How is text size set for rendered text?

File: ui/renderers/html.go [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+)
+
+type HTMLRenderer struct {
+	doc       ast.Node
+	container *layout.List
+	*Renderer
+}
+
+var (
+	blockEls = []string{"div", "p", "h1", "h2", "h3", "h4", "h5", "h6", "ul"}
```
Comment (beansgum) [inline ui/renderers/html.go]: There's no point having a `<ul>` tag without `<ol>` and `<li>`.

File: ui/renderers/renderer.go [hunk 11]
```diff
@@ -148,11 +110,20 @@ func (r *Renderer) RenderNode(w io.Writer, node ast.Node, entering bool) ast.Wal
...
+	r.renderWords(r.theme.Body2(""))
 	// add dummy widget for new line
 	r.renderEmptyLine()
 }
 
+func (r *Renderer) renderEmph() {
+	label := r.theme.Body2("")
+	label.Font.Style = text.Italic
+
+	fmt.Println("ssss")
```
Comment (beansgum) [inline ui/renderers/renderer.go]: 🐍

**Review by beansgum — CHANGES_REQUESTED** [review]
The text size doesn't look correct


- Attempting to render a font tag produces this
```html
<font color="danger">INVALID</font>
```


- Line break does not work
```html
<span style="text-color: primary">INVALID</span> <br/> <span style="text-color: gray">INVALID</span>
```


- The styling should end where the tag ends.
```html
Tap on <span style="text-color: primary">blue text</span> to copy the item.
```

**Review by beansgum — CHANGES_REQUESTED** [review]
- Color codes don't work in the font tag
```html
<font color="#2970FF">Confirmed</font> &centerdot; 6 Confirmations
```


- styling does not recognize the color attribute
```html
<span style="color: gray">
    Signing a message with an address' private key allows you to prove that 
    you are the owner of a given address to a possible counterparty.
</span>
```

**Review by beansgum — CHANGES_REQUESTED** [review]
- The renderer adds an empty space after a tag
```html
<font color="#ed6d47">INVALID</font>.
```

**Review by beansgum — CHANGES_REQUESTED** [review]
- The renderer removes empty space included space after a tag
```html
<font color="#ed6d47">INVALID</font> Space.
```

**Review by beansgum — CHANGES_REQUESTED** [review]
The renderer removes empty space included space before a tag.

```html
otherwise <font color="#ed6d47">INVALID</font>.
```

**Review by beansgum — APPROVED** [review]
Header tags are supposed to render big and bold but we don't need those.
```html
<h1>This is heading 1</h1>
<h2>This is heading 2</h2>
<h3>This is heading 3</h3>			
<p>This is a paragraph.</p>
<p>This is a <br/> paragraph with a line break.</p>
<p style="color: #FF0000;">This is a red paragraph.</p>
<p><b>This text is bold</b></p>
<p><i>This text is italic</i></p>
<strong>This text is important!</strong>
```

**Review by oshorefueled — APPROVED** [review]

**Pull request merged**